### PR TITLE
BTS-693 fix for limit with materializer

### PR DESCRIPTION
**Pull request by Dronplane**
### Scope & Purpose

Fix call limits for the constrained sort node by placing additional limit node right after the constrained sort (or moving existing node if possible)

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] Backport for 3.9: *(Please link PR)*
- [ ] Backport for 3.8: *(Please link PR)*
- [ ] Backport for 3.7: *(Please link PR)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

**Review by gnusi — APPROVED** [review]
LGTM

File: arangod/Aql/OptimizerRule.h [hunk 1]
```diff
@@ -358,6 +358,13 @@ struct OptimizerRule {
 
   static_assert(scatterInClusterRule < parallelizeGatherRule);
 
+  static_assert(moveCalculationsUpRule < applySortLimitRule,
```
Comment (mchacki) [inline arangod/Aql/OptimizerRule.h:361]: Very good :+1:

**Review by mchacki — CHANGES_REQUESTED** [review]
Technical solution looks very good :+1:

Requested Change: We need some tests to back this up, the issue triggered here is rather tricky so needs a good coverage.

File: arangod/Aql/ExecutionEngine.cpp [hunk 2]
```diff
@@ -52,6 +52,71 @@
...
+// 3. Between the “root()” of the Plan (which in most cases is the RETURN
+//    Node) and the LIMIT with fullCount:true no other LIMIT is allowed
+//    (exception the ContrainedSort case now)
+struct LimitFullCountChecker final
+    : public WalkerWorker<ExecutionNode, WalkerUniqueness::Unique> {
+  size_t subqueryRecursionCounter{0};
+  bool seenFullCount{false};
+  bool seenFullCountLimitWithoutSort{false};
+
+  explicit LimitFullCountChecker() noexcept {}
```
Comment (gnusi) [inline arangod/Aql/ExecutionEngine.cpp:73]:
```suggestion
LimitFullCountChecker() = default;
```

**Review by gnusi — APPROVED** [review]

File: arangod/Aql/ExecutionEngine.cpp [hunk 3]
```diff
@@ -52,6 +52,73 @@
...
+using namespace arangodb::aql;
+
+// Validating fullCount usage.
+// For fullCount the following applies:
+// 1. There can at most be one LIMIT that has fullCount: true
+// 2. Limit with fullCount:true cannot be within a subquery.
+// 3. Between the “root()” of the Plan (which in most cases is the RETURN
+//    Node) and the LIMIT with fullCount:true no other LIMIT is allowed
+//    (exception the ContrainedSort case now)
+struct LimitFullCountChecker final
```
Comment (mchacki) [inline arangod/Aql/ExecutionEngine.cpp:67]: This looks very good, thanks for implementing :+1:

File: tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch.inc [hunk 4]
```diff
@@ -407,6 +439,60 @@
           expected.delete(doc.key);
         });
         assertEqual(0, expected.size);
+      },
+      // fullCount was too low
+      testRegressionBts611() {
```
Comment (mchacki) [inline tests/js/server/aql/aql-optimizer-rule-late-document-materialization-arangosearch.inc:444]: Like this test 👍

**Review by mchacki — APPROVED** [review]
LGTM

**Pull request merged**